flang-new -> flang renaming article (#62)
llvm/llvm-blog-www: content/posts/2024-11-05-flang-new.md

---
author: "David Spickett"
date: "2024-11-05"
tags: ["Flang", "MLIR"]
title: "LLVM Fortran Levels Up: Goodbye flang-new, hello flang!"
---

LLVM has included a Fortran compiler "Flang" since LLVM 11. What you might not

> know is that the name of Flang's binary was not `flang`, but `flang-new`.

Reviewer: Note "LLVM Flang" includes multiple binaries, `flang` (fka `flang-new`) happens to be Flang's compiler driver. However, when "LLVM Flang" was merged into LLVM, `flang` was not a binary ;-)

Author: I will find a way to reword this such that it does not imply that *in llvm 11* there was a flang-new, but that just prior to llvm 20 you would have found only flang-new.

Author: Done.

The LLVM 20 release brings an end to the era of `flang-new`. The
community has decided Flang has reached a level of quality that is worthy of a
new name.

The "new" name? You guessed it, `flang`.

In this article I will explain why this simple change actually represents a
major milestone for the Flang project. Covering Flang's almost 10 year journey to

> into the LLVM Project and its adoption of one of LLVM's hottest new features, MLIR.

Reviewer: It's 10 yrs if you consider "Flang" == "LLVM Flang" | "Classic Flang". I've been trying to make clear distinction to avoid confusion. Ultimately, these are two completely different projects. And some (most?) commercial are based on "Classic" rather than "LLVM" Flang.

Author: In this context I'm referring to the genesis of an open source LLVM based Fortran compiler that in some way got named flang. So 10 years since the National Labs announcement.

But I can be clearer here, if it didn't make sense to the reader then it needs to change regardless.

[Intel Fortran Compiler](https://www.intel.com/content/www/us/en/developer/tools/oneapi/fortran-compiler.html)
or NVIDIA's [HPC compilers](https://developer.nvidia.com/hpc-compilers) and
there are open source options like [gfortran](https://gcc.gnu.org/fortran/).

So why build another one? In 2020 Gary Klimowicz (NVIDIA)
[laid out](https://www.youtube.com/watch?v=Fy68k5hHgLk) the goals for Flang,
paraphrased here:

* Develop Flang under a permissive license like LLVM's, which is more palatable
  to commercial users and contributors.
* Develop an active community of Fortran compiler developers that includes
  companies and institutions.
* Support Fortran tool development by being based on existing LLVM frameworks.
* Support Fortran language experimentation for future language standards proposals.

The last one can be explained by comparison to Clang. For C++ feature, a common
path is to implement them on top of Clang, GCC, or both, to prove the feature can work.
Implementing the feature multiple times in different compilers uncovers any
assumptions in the design that may be a problem for one compiler or another.

> In the same way, Flang and gfortran can provide that diversity.

Reviewer: ```suggestion
In the same way, Flang and GFortran can provide that diversity.
```

Author: It would help if GNU would stick to their own branding (https://gcc.gnu.org/wiki/GFortran) but yes, I will fix all the instances of this.

Author: Fixed.

  STMicroElectronics.
* **July 2013** PGI is [sold to NVIDIA](https://www.theregister.com/2013/07/30/nvidia_buys_the_portland_group/).

In late 2015 we saw the first signs of what would become "Classic Flang".

* **November 2015** NVIDIA is part of the US Department of Energy
  Exascale Computing Project. Including a commitment to create an open source
  [Fortran compiler](https://www.llnl.gov/article/41756/nnsa-national-labs-team-nvidia-develop-open-source-fortran-compiler-technology).

  > “The U.S. Department of Energy’s National Nuclear Security Administration and its
  > three national labs have reached an agreement with NVIDIA's PGI division to adapt
  > and open-source PGI's Fortran frontend, and associated Fortran runtime library,
  > for contribution to the LLVM project.”
* **May 2017** The first source code release of Classic Flang as a separate
  respository, outside of the LLVM Project. Consisting of PGI's compiler frontend
  and a new backend that generated LLVM IR.

  Other companies such as Arm and AMD got involved at this point.

During this time, plans were being made to propose Classic Flang for inclusion

> in the LLVM Project.

Reviewer: You may want to explain somewhere that _today_ we make the distinction into "Classic Flang" and "LLVM Flang", but back then it was just ... Flang.

Author: I've noted this in the prior section where I first use the term, and that I will use classic flang even when classic flang wasn't a term yet, for clarity's sake.

Around this time work started in the Classic Flang repo on the `fir-dev` branch
that would enable code generation when using F18.

So for the next events remember that this whole time Classic Flang was (and still
is today) in use. Classic Flang's binary name was `flang`, just like the folder F18 now occupied
in the LLVM monorepo.

* **October 2020** Flang is included in an LLVM release for the first time in
  LLVM 11.0.0. There was an `f18` binary and the aforementioned script
  `flang.sh`.
* **October 2021** First appearance of the `flang-new` binary, in the LLVM 13.0.0
  release.
* **April 2022** NVIDIA [ceases development](https://discourse.llvm.org/t/nvidia-transition-from-fir-dev/61947)
  of the `fir-dev` branch in the Classic Flang project. Upstreaming of `fir-dev`
  to the LLVM Project begins around this date.
* **April 2022** `flang-new` can now do code generation if an experimental flag
  is used. Enabled by the code upstreamed into the LLVM Project from `fir-dev`.
* **May 2022** The `flang.sh` driver script becomes `flang-to-external-fc`. It
  allows the user to use `flang-new` to process Fortran source code then send

>   the result to an existing Fortran compiler such as gfortran.

Reviewer: "process" -> "parse and unparse" (if you want to be specific)

Author: Done.

  (this seems strange, but it allowed users to drop in Flang as a replacement
  for an existing compiler and test Flang's parsing code on their own codebases)
* **June 2022** `flang-new` was changed to do code generation
  [by default](https://github.com/llvm/llvm-project/commit/0867d4157328169c570f4d1e9c01806624035873)
  with no extra flags required.
* **July 2022** NVIDIA [finish](https://discourse.llvm.org/t/nvidia-transition-from-fir-dev/61947/5)

>   upstreaming `fir-dev`.

Reviewer: The upstreaming was carried out by multiple members of the community, not only NVIDIA :)

Author: Fixed.

I am talking to some folks about FIR so I will no doubt be able to correct this and other things soon.

An initiative to upstream Classic Flang had effectively rewritten the entire
compiler by the time it was complete.

<!-- For some reason Hugo doesn't render any plaintext in these bullet points
     correctly unless I they have blank lines between them. -->

* **April 2024** The `flang-to-external-fc` script is removed. Marking the point
  where `flang-new` was mature enough to be used as a standalone compiler.

* **October 2024** The community deems that Flang has met the criteria to not be
  "new" and the name is changed. Goodbye `flang-new`, hello `flang`!

* TODO: First release with `flang` binary goes here!

# Flang and the Definition of New

The renaming of Flang was a contentious
[proposal](https://discourse.llvm.org/t/proposal-rename-flang-new-to-flang/69462)
and an early use of the
[LLVM proposal process](https://github.com/llvm/llvm-www/blob/main/proposals/LP0001-LLVMDecisionMaking.md).

The most critial point was that `flang` has meant different things to different
people over time.

To some `flang` was the out of tree Classic Flang. To some it was the in-tree
LLVM Flang. Some were aware of both, but really wanted to move on from Classic

> Flang.

Reviewer: IMHO, it would be good to introduce "LLVM Flang" (with quotes) somewhere in the text, similarly to how "Classic Flang" is introduced.

Main rationale - once published, people will likely use this blog post as their main reference for what "Flang" actually is. So this might be the only opportunity to clarify the distinction and to avoid confusion.

As for "Flang" being 10 yrs old ... I see where you're coming from, makes sense 👍🏻

Author: Yes this is a good benchmark. Would this article be a good (albeit long) answer to "which flang is/was which".

Author: I've improved the place where I first use the term.

> implementation challenges—by making it very cheap to define and introduce new
> abstraction levels, and provide “in the box” infrastructure to solve common
> compiler engineering problems.
- ["MLIR: A Compiler Infrastructure for the End of Moore’s Law"](https://arxiv.org/abs/2002.11054)
  (Chris Lattner et al., 2020)

# Flang and MLIR

The same year MLIR debuted, Eric Schweitz (NVIDIA) gave a talk at the later US
LLVM Developers' meeting titled
["An MLIR Dialect for High-Level Optimization of Fortran”](https://www.youtube.com/watch?v=ff3ngdvUang).

Remember the `fir-dev` branch? The `fir` is this.
[Fortran Intermediate Representation](https://flang.llvm.org/docs/FIRLangRef.html),
better known to developers as "FIR".

FIR did the same thing that Rust, Swift and Julia did but this time, Flang had
the benefit of basing FIR on MLIR.

They did not stop there, later adding
["High Level Fortran Intermediate Representation"](https://flang.llvm.org/docs/HighLevelFIR.html)
(HLFIR) which works at a higher level than FIR. A big target of HLFIR 
was array optimisations, which were more complex to handle using just FIR.

## IRs All the Way Down

> We start with Fortran source code. This function assigns array `b` to array `a`.

Author: Fixed.

```fortran
subroutine foo(a, b)
  real :: a(:), b(:)
  a = b
end subroutine
```

(if you want to modify this example, you can do so on
[Compiler Explorer](https://godbolt.org/z/851qrcj6a))

It is tempting to think of the IRs in a "stack" where each one is converted
into the next. However, MLIR allows multiple "dialects" of MLIR to exist in the
same file.

As you can see in this first step, Flang produces a file that uses the HLFIR

> dialect where it can, and the FIR dialect for the rest.

Author: I didn't notice that, mentioning it only furthers the point about dialects co-existing so I've done that.

```mlir
module attributes {<...>} {
  func.func @_QPfoo(%arg0: !fir.box<!fir.array<?xf32>> {fir.bindc_name = "a"}, %arg1: !fir.box<!fir.array<?xf32>> {fir.bindc_name = "b"}) {
    %0 = fir.dummy_scope : !fir.dscope
    %1:2 = hlfir.declare %arg0 dummy_scope %0 {uniq_name = "_QFfooEa"} : (!fir.box<!fir.array<?xf32>>, !fir.dscope) -> (!fir.box<!fir.array<?xf32>>, !fir.box<!fir.array<?xf32>>)
    %2:2 = hlfir.declare %arg1 dummy_scope %0 {uniq_name = "_QFfooEb"} : (!fir.box<!fir.array<?xf32>>, !fir.dscope) -> (!fir.box<!fir.array<?xf32>>, !fir.box<!fir.array<?xf32>>)
    hlfir.assign %2#0 to %1#0 : !fir.box<!fir.array<?xf32>>, !fir.box<!fir.array<?xf32>>
    return
  }
}
```

> For example, the parameters are declared with `hlfir.declare` but their type is

Reviewer 2: If you want to use Fortran terminology here, `parameters` must be replaced by `dummy arguments`.

Author: I've used the term and linked to FortranLang's explanation of it, to save space.

  ret void
}
<...>
```
This LLVM IR passes through the standard optimisation pipeline that LLVM IR produced
by Clang would. Eventually being converted into target specific Machine IR
(MIR) and finally machine code.

At each of the stages, the optimisations most suited to that stage are done.
For example, while we are in HLFIR we could optimise array accesses because in
HLFIR we have information about how the Fortran language treats those.

If we were to do this later in LLVM IR it would be much more difficult. Either
the information would be lost or incomplete, or we would be at a stage in the pipeline
where we cannot assume that we started with a specific source language.

# Flang Takes Driving Lessons

**Note:** This section paraphrases material from
["Flang Drivers"](https://github.com/llvm/llvm-project/blob/main/flang/docs/FlangDriver.md).
If you want more detail please refer to that document.

"Driver" in a compiler context means the part of the compiler that decides
how to handle a set of options. For instance, when you use the option `-march=armv8a+memtag`,
something in Flang knows that you want to compile for Armv8.0-a with the memory

> tagging extension enabled.

Reviewer: You can add a link to:
* https://fabiensanglard.net/dc/index.php

Author: Added it in the initial note.

`-march=` is an example of a "compiler driver" option. These options are what users
give to the compiler. There is actually a second driver after this, confusingly
called the "frontend" driver, despite being behind the scenes.

In this case the "compiler driver" is `flang` and the "frontend driver" is

> `flang -fc1` (they are in the same binary).

Reviewer: Perhaps less of "they are in the same binary" and more of "these are effectively two separate tools accessed through one binary".

I don't know the full story behind this split in Clang (and GCC?), but from my experience:
* not having a dedicated frontend driver binary makes it easier to prevent people accidentally using a low level tool
* (Clang/Flang specific) having one binary means re-using the logic for defining/parsing options (there's a huge overlap between `flang` and `flang -fc1`),
* no need to start a new process to run the frontend driver.

Author: Done.

The main reason they are split into 2 is so that the "compiler driver" can have
option names that do not change over time and the "frontend driver" can change

> its option names as suits the developers of the compiler.

Reviewer: Another argument (well, same argument, different phrasing):
* compiler driver is for end-users who want to generate executables,
* frontend driver is for compiler developers interested in the internals of the Frontend/Flang itself.

Author: I've gone with a mix of both.

arguments to `flang -fc1`. This means that if compiler developers decided to
change how `-march=` was converted, existing compile commands for `flang` would
still work.

Another compiler driver responsibility is to know where to find libraries on various
systems. This differs between operating systems and even distributions of the
same family of operating systems (for example Linux distributions).

This created a problem when implementing Flang's compiler driver. This many
details would take a long time to get right.

Luckily for Flang, by this time it was in the LLVM Project, alongside Clang.
Clang already knew how to handle this and had been tested on all sorts of
systems over many years.

> The intent is to mirror clang, for both the driver and CompilerInvocation, as
> much as makes sense to do so. The aim is to avoid re-inventing the wheel and
> to enable people who have worked with either the clang or flang entrypoints,
> drivers, and frontends to easily understand the other.
* [Peter Waller](https://discourse.llvm.org/t/rfc-adding-a-fortran-mode-to-the-clang-driver-for-flang/52307) (Arm)

Flang became the the first project to use Clang's compiler driver library

> (`clangDriver`) to implement its own compiler driver.

Reviewer: Welll ... IIRC, "Classic Flang" did the same, but that was out-of-tree. There might be other out-of-tree examples. But "LLVM Flang" remains the only in-tree example.

Author: I will note that, but given that Classic Flang is in the same lineage I think I can make almost the same statement still.

Author: I've said "first in tree".

Doing so:
* Gave Flang support for a large number of targets and operating systems.
* Let Flang control any linker or assembler that Clang could.

> * Made it possible to add any Clang option to Flang.

Reviewer: I would avoid this one. It creates an impression that any option available in Clang should auto-magically be available in Flang. That's not always the case. "LLVM" options are a good example (e.g. `-fpie`) - these need to be "wired-up" correctly so that the interaction between them is correct and the "LLVM" backend is ... driven accordingly. 

But as far as plain option parsing is concerned, Flang can leverage all the greatness from Clang, yes.

Author: I did note afterwards the caveat, but removing it would let me delete that note and it's not super important anyway. So I will do this.

Author: Removed the options part completely.

by Clang would. Eventually being converted into target specific Machine IR
(MIR) and finally machine code.

At each of the stages, the optimisations most suited to that stage are done.
For example, while we are in HLFIR we could optimise array accesses because in
HLFIR we have information about how the Fortran language treats those.

If we were to do this later in LLVM IR it would be much more difficult. Either
the information would be lost or incomplete, or we would be at a stage in the pipeline
where we cannot assume that we started with a specific source language.

# OpenMP to Everyone

**Note:** Most of the points made in this section also apply to [OpenACC](https://www.openacc.org/) support for Flang. In the interest of time, I will only talk
about OpenMP in this article. You can find out more about OpenACC support in this
[presentation](https://www.youtube.com/watch?v=vVmCLdSboWc).

## The Basics

OpenMP is a standardised API for adding parallelisation to C, C++ and
Fortran programs. Programmers mark parts of their code with directives.
The compiler interprets these directives and produces calls to a runtime library

> that distributes work as instructed.

Author: It is misleading yes, I've replaced it with something akin to your version.

Here is an example in Fortran:
```Fortran
SUBROUTINE SIMPLE(N, A, B)
  INTEGER I, N
  REAL B(N), A(N)
!$OMP PARALLEL DO
  DO I=2,N
    B(I) = (A(I) + A(I-1)) / 2.0
  ENDDO
END SUBROUTINE SIMPLE
```
(taken from ["OpenMP Application Programming Interface Examples"](https://www.openmp.org/wp-content/uploads/openmp-examples-4.5.0.pdf), [Compiler Explorer](https://godbolt.org/z/chjzs3o6r))

**Note:** Fortran arrays are [one-based](https://fortran-lang.org/en/learn/quickstart/arrays_strings/) by default. So the first element is at index 1. This example reads the previous element, so it starts `I` at 2.

> `!$OMP PARALLEL DO` is a directive in the form of a Fortran comment, which

Reviewer 2: ```suggestion
`!$OMP PARALLEL DO` is a combined directive in the form of a Fortran comment, which
```

Author: A combined directive is a subset of directives is what I'd argue here. From a reader's point of view not the OpenMP standard's perspective.

Unless you think not including that will make it misleading.

Reviewer 2: Agreed.

start with `!`. `PARALLEL DO` starts a parallel "region" which includes the
code from `DO` to `ENDDO`.

This tells the compiler that the work in the `DO` loop should be shared amongst
all the threads available to the program (the number of threads is controlled
by settings in the program's execution environment).

Clang has [supported OpenMP](https://blog.llvm.org/2015/05/openmp-support_22.html)
for many years now. The equivalent C++ is:
```C++
void simple(int n, float *a, float *b)
{
    int i;

    #pragma omp parallel for
    for (i=1; i<n; i++)
        b[i] = (a[i] + a[i-1]) / 2.0;
}
```
([Compiler Explorer](https://godbolt.org/z/eca6vjbqe))

> Instead of `!`, comments start with `#` and instead of `DO` there is `for`.

Author: Must've thought I was in Python, fixed.

Overall very similar to the Fortran example.

LLVM IR does not know anything about OpenMP specifically, so Clang uses a

> library called `OpenMPIRBuilder` to convert OpenMP directives to LLVM IR.

Author: I've removed the mention here and instead there is a small section on the effort to move that code out of Clang later.

```mlir
define dso_local void @simple(<...>) {
<...>
  call void <...> @__kmpc_fork_call(<...> ptr @simple(<...>) (.omp_outlined), <...>
  ret void
}

> define internal void @simple(<...>) (.omp_outlined) <...>) {

Author: Fixed, and elsewhere.

<...>
  call void @simple(<...>) (.omp_outlined_debug__) <...>
  ret void
}

define internal void @simple(<...>) (.omp_outlined_debug__) <...> {
<...>
  call void @__kmpc_for_static_init_4(ptr @1, i32 %8, i32 34, <...>)
<...>
omp.inner.for.body:
<...>
omp.precond.end:
  ret void
}
```

`simple` no longer does the work of the program. Instead it uses
`__kmpc_fork_call` to call another version of `simple`, which calls yet another

> version. `__kmpc_fork_call` is part of the OpenMP

Author: I've added one, please review it.

> version. `__kmpc_fork_call` is part of the OpenMP

Reviewer 2: LGTM, thanks for adding it.

They are just normal LLVM IR labels.

## Bringing OpenMP to MLIR

Early in 2020, Kiran Chandramohan (Arm) [proposed](https://discourse.llvm.org/t/rfc-openmp-dialect-in-mlir/397) an MLIR dialect for OpenMP, with the end goal
of Flang supporting OpenMP.

> We (Arm) started the work for the OpenMP MLIR dialect because of Flang.
> ... So, MLIR has an OpenMP dialect because of Flang.
* Kiran Chandramohan

This dialect would represent OpenMP specifically, unlike the generic LLVM IR you
get from Clang.

If you go back to the Fortran OpenMP example and compile it without OpenMP
enabled, you get this MLIR:

```mlir
module attributes {<...>} {
  func.func @_QPsimple(<...> {
    %1:2 = hlfir.declare %arg0 <...> {uniq_name = "_QFsimpleEn"} : <...>
    %3:2 = hlfir.declare %2 <...> {uniq_name = "_QFsimpleEi"} : <...>
    %10:2 = hlfir.declare %arg1(%9) <...> {uniq_name = "_QFsimpleEa"} : <...>
    %17:2 = hlfir.declare %arg2(%16) <...> {uniq_name = "_QFsimpleEb"} : <...>
    %22:2 = fir.do_loop <...> {
      <...>
      hlfir.assign %34 to %37 : f32, !fir.ref<f32>
    }
    fir.store %22#1 to %3#1 : !fir.ref<i32>
    return
  }
}
```
(output heavily edited for readability)

> Notice that the `DO` loop has been converted into `for.do_loop`. Now enable

Author: Fixed.

OpenMP and compile again:

```mlir
module attributes {<...>} {
  func.func @_QPsimple(<...> {
    %1:2 = hlfir.declare %arg0 <...> {uniq_name = "_QFsimpleEn"} : <...>
    %10:2 = hlfir.declare %arg1(%9) <...> {uniq_name = "_QFsimpleEa"} : <...>
    %17:2 = hlfir.declare %arg2(%16) <...> {uniq_name = "_QFsimpleEb"} : <...>
    omp.parallel {
      %19:2 = hlfir.declare %18 {uniq_name = "_QFsimpleEi"} : <...>
      omp.wsloop {
        omp.loop_nest (%arg3) : i32 = (%c2_i32) to (%20) inclusive step (%c1_i32) {
          hlfir.assign %32 to %35 : f32, !fir.ref<f32>
          omp.yield
        }
      }
      omp.terminator
    }
    return
  }
}
```
(output heavily edited for readability)

> You will see that instead of `for.do_loop` you have `omp.parallel`. `omp` is

Author: Agreed, done.

MLIR's [OpenMP dialect](https://mlir.llvm.org/docs/Dialects/OpenMPDialect/).
This is a higher level and much more literal translation of the

> `parallel` directive than Clang produced.

Reviewer 2: ```suggestion
`parallel do` combined directive than Clang produced.
```

Author: directive / combined directive, it's the same thing to the casual reader in my opinion. So I will leave it as is unless it's misleading?

I have refactored this bit some so the wording is not exactly the same. I tried to make it clear that Clang and Flang aren't processing literally the same directive here, but the intent is the same.

Reviewer 2: Right, the distinction between directive / combined directive should not matter to the casual reader.

I saw that in the latest version of the text you have used `parallel do` and `parallel for` instead, so I think it's fine now.

As the `omp` dialect is specifically made for OpenMP, it can represent
it in a much more natural manner. This makes understanding the code and writing
optimisations much easier.

Of course Flang needs to produce LLVM IR eventually, and it reused Clang's
`OpenMPIRBuilder` library for this. Below is the LLVM IR produced from the MLIR
above:

```mlir

> define void @simple_(<...> {

Reviewer 2: nit: this is a matter of preference (so feel free to ignore it), but the omitted closing parentheses in this snippet makes it harder for me to parse it.

Author: Yeah I've cleaned up a few of these. It'll help because there is formatting for MLIR so it'll colour matched sets nicely.

entry:
  call void (ptr, i32, ptr, ...) @__kmpc_fork_call( <...> @simple_..omp_par <...>
  ret void
}

define internal void @simple_..omp_par( <...> {
omp.par.entry:
  call void @__kmpc_for_static_init_4u( <...>
omp_loop.exit:
  call void @__kmpc_barrier(ptr nonnull @2, i32 %omp_global_thread_num4)
  ret void
omp_loop.body:
  <...>
}
```

> There are some differences between this and the IR Clang produced but

Reviewer 2: Adding `-O3` to Clang example reduces the differences.

Author: Yes it did thankyou, I've done that.

Still cosmetic differences so I've emphasised the structural similarity.

this is not unexpected. The original source language is different and Flang's
IR also goes through MLIR's compiler passes.

## ClangIR and the Future

> It is surprising that a compiler for a language as old as Fotran got ahead of

Reviewer 2: ```suggestion
It is surprising that a compiler for a language as old as Fortran got ahead of
```

Author: Fixed.

the micro task function is called, it gets a different slice of the work.

The number of instances is only known at runtime, and can be controlled with
settings such as [`OMP_NUM_THREADS`](https://www.openmp.org/spec-html/5.0/openmpse50.html).

This function's LLVM IR includes labels like `omp.loop.exit.i` but these are
not specific to OpenMP. They are just normal LLVM IR labels whose name includes
`omp`.

## Sharing Clang's OpenMP Knowledge

In April 2019 LLVM Flang ("F18" at the time) was approved to join the LLVM
Project. A month later, it was proposed that the soon-to-be LLVM Flang should
leverage Clang's knowledge of OpenMP.

> This is an RFC for the design of the OpenMP front-ends under the LLVM
> umbrella. It is necessary to talk about this now as Flang (aka. F18) is
> maturing at a very promising rate and about to become a sub-project next
> to Clang.
>
> TLDR;
> Keep AST nodes and Sema separated but unify LLVM-IR generation for
> OpenMP constructs based on the (almost) identical OpenMP directive
> level.
- "[RFC] Proposed interplay of Clang & Flang & LLVM wrt. OpenMP",
  Johannes Doerfert, May 2019 (from a copy provided to me, no copies appear

>   to exist online at this time)

Reviewer 3: Here's my small contribution: the RFC is still available. You can include it for completeness: [RFC: Proposed Interplay of Clang, Flang, LLVM wrt OpenMP](https://discourse.llvm.org/t/rfc-proposed-interplay-of-clang-flang-llvm-wrt-openmp-flang-dev/51905).

Author: Done.

* **September 2020** Flang's new compiler driver is [added](https://reviews.llvm.org/D86089)
  as an experimental option. This is the first appearance of the `flang-new` binary,
  instead of `flang-tmp` as proposed before.

  > The name was intended as temporary, but not the driver.
  * Andrzej Warzyński (Arm, Flang Driver Maintainer)

* **October 2020** Flang is included in an LLVM release for the first time in
  LLVM 11.0.0. There was an `f18` binary and the aforementioned script
  `flang.sh`.
* **August 2021** `flang-new` is no longer experimental and [replaces](https://reviews.llvm.org/D105811)
  the previous Flang compiler driver `f18`.
* **October 2021** First appearance of the `flang-new` binary in an LLVM release
 (13.0.0).
* **April 2022** NVIDIA [ceases development](https://discourse.llvm.org/t/nvidia-transition-from-fir-dev/61947)
  of the `fir-dev` branch in the Classic Flang project. Upstreaming of `fir-dev`
  to the LLVM Project begins around this date.

  `flang-new` can now do [code generation](https://reviews.llvm.org/D122008)
  if the `-flang-experimental-exec` option is used. This change leveraged work

>   originally done on the `fir-dev` branch.

Reviewer 4: It would be nice to note here (and below) which LLVM version the various milestones first appeared in.

Author: Yes good idea. I have added the significant releases as their own entries in the timeline.

It's good because now we can see that (at least from an external perspective) there was only a month between the first release with `flang-new` as the only Flang binary, and the decision to rename it to `flang`.

* **May 2022** Kiran Chandramohan (Arm) [announces](https://www.youtube.com/watch?v=FoIjafZGDdE)

>   at th European LLVM Developers' Meeting that Flang's OpenMP 1.1 support is close to complete.

Reviewer 4: ```suggestion
  at the European LLVM Developers' Meeting that Flang's OpenMP 1.1 support is close to complete.
```

Author: Fixed.

that it was not yet time to change the name.

So the community got to work on those goals:

* Many performance and correctness issues were addressed by the work to implement "High Level Fortran Intermediate Representation" (HLFIR).
* A cross-company team including Arm, Huawei, Linaro, Nvidia and Qualcomm
  [collaborated](https://github.com/orgs/llvm/projects/35/views/1)
  to make it possible to build the popular [SPEC 2017](https://www.spec.org/cpu2017/)
  benchmark with Flang.
* Flang gained support for OpenMP up to version 2.5, and was able to compile OpenMP
  specific benchmarks like [SPEC OMP](https://www.spec.org/omp2012/) and the
  [NAS Parallel Benchmarks](https://www.nas.nasa.gov/software/npb.html).
* Linaro [showed that](https://www.youtube.com/watch?v=Gua80XRPhyY) the performance
  of Flang compared favourably with Classic Flang and was not far behind GFortran.
* The GFortran test suite was added to the [LLVM Test Suite](https://github.com/llvm/llvm-test-suite/tree/main/Fortran/gfortran),
  and Flang achieved good results.
* Test suites from [Fujitsu](https://github.com/fujitsu/compiler-test-suite) and

>   IBM were made public and tested with Flang.

Author: Ok, I know there is some public repo but I didn't link to it because I was told it was not fully ready.

I have relaxed the language here, but still included it because I think it is/will be an important contribution.

With all that done, in October of 2024 `flang-new`
[became](https://github.com/llvm/llvm-project/commit/06eb10dadfaeaadc5d0d95d38bea4bfb5253e077)
`flang`. The future had arrived.

> And it’s merged! It’s been a long (and sometimes contentious) process, but
> thank you to everyone who contributed to the discussion.
- Brad Richardson, [closing out](https://discourse.llvm.org/t/proposal-rename-flang-new-to-flang/69462/86) the proposal.

The goals the community achieved have certainly been worth it for Flang as a
compiler, but did Brad achieve their own goals?

> What did I hope to see as a result of the name change? I wanted it to be
> easier for more people to try it out.

So once you have finished reading this article, download Flang or try it out
on Compiler Explorer. You know at least one person will appreciate it!

TODO: download link goes here
TODO: CE link goes here.

# Fortran Intermediate Representation (FIR)

All compilers that use LLVM as a backend eventually produce code in the form of
the [LLVM Intermediate Representation](https://llvm.org/docs/LangRef.html)
(LLVM IR).

> A drawback of this is that LLVM IR does not include language specific information
> . This means that for example, it cannot be used to optimise arrays in a way

Reviewer 4: ```suggestion
A drawback of this is that LLVM IR does not include language specific information.
This means that for example, it cannot be used to optimise arrays in a way
```

Author: Fixed.

specific to Fortran programs.

One solution to this has been to build a higher level IR that represents the
unique features of the language, optimise that, then convert the result into LLVM IR.

Eric Schweitz started to do that for Fortran in late 2018:

> FIR was originally conceived as a high-level IR that would interoperate with
> LLVM but have a representation more friendly and amenable to Fortran
> optimizations.

Naming is hard but Eric defied expectations:

> FIR was a pun of sorts. Fortran IR and meant to be evocative of the trees
> (Abstract Syntax Trees).

We will not go into detail about this early FIR, because [MLIR](https://mlir.llvm.org/)
was revealed soon after Eric started the project and they quickly adopted it.

> When MLIR was announced, I quickly switched gears from building data
> structures for a new "intermediate IR" to porting my IR design to MLIR and
> using that instead.
>
> I believe FIR was probably the first "serious project" outside of Google to
> start using MLIR.

The FIR work continued to develop, with Jean Perier (Nvidia) joining Eric on

> the project. It became its own public branch `fir-dev`, that was latercontributed to the LLVM Project.

Reviewer 4: ```suggestion
the project. It became its own public branch `fir-dev`, that was later contributed to the LLVM Project.
```

Author: Fixed.

The following sections will go into detail on the IRs that Flang uses today.

# MLIR

The journey from Classic Flang to LLVM Flang involved a rewrite of the
entire compiler. This provided an opportunity to pick up new things from
the LLVM Project. Most notably MLIR.

"Multi-Level Intermediate Representation" (MLIR) was first
[introduced](https://www.youtube.com/watch?v=qzljG6DKgic) to the LLVM
community in 2019 (around the time that F18 was approved to move into the LLVM Project).

> The problem that MLIR addresses is the same one that Eric Schweitz tackled with FIR.

Reviewer 4: ```suggestion
The problem that MLIR addresses is the same one that Eric Schweitz tackled with FIR:
```

Author: Done.

It is difficult to map high level details of programming languages
into LLVM IR.

You either have to to attach them to the IR as metadata, try to recover the
lost details later, or fight an uphill battle to add the details to
LLVM IR itself. These details are crucial for producing optimised code in certain
languages (Fortran array optimisations were an early cited use case for MLIR).

> This lead languages such as Swift and Rust to create their own IRs that include

Reviewer 5: ```suggestion
This led languages such as Swift and Rust to create their own IRs that include
```

Author: Fixed, also found another instance of lead -> led, it's one of my blind spots clearly.

> It contributed significantly to establishing and validating that MLIR had the right foundations.
- Mehdi Amini

Flang did not stop there, later adding another dialect
["High Level Fortran Intermediate Representation"](https://flang.llvm.org/docs/HighLevelFIR.html)
(HLFIR) which works at a higher level than FIR. A big target of HLFIR
was array optimisations, that were more complex to handle using FIR alone.

> FIR was a compromise on both ends to some degree. It wasn't trying to capture
> syntactic information from Fortran, and I assumed there would be work done on
> an Abstract Syntax Tree. That niche would later be filled by "High Level FIR"
> [HLFIR].
- Eric Schweitz

## IRs All the Way Down

The compilation process starts with Fortran source code.

```fortran
subroutine example(a, b)
  real :: a(:), b(:)
  a = b
end subroutine
```
([Compiler Explorer](https://godbolt.org/z/8j3W46j3j))

The
[subroutine](https://fortran-lang.org/learn/quickstart/organising_code/) `example`
assigns array `b` to array `a`.

It is tempting to think of the IRs in a "stack" where each one is converted
into the next. However, MLIR allows multiple "dialects" of MLIR to exist in the
same file.

> (the steps shown here are the most important ones for Flang, in reality there
> are many more between Fortran and LLVM IR)

Reviewer 5: ```suggestion
(The steps shown here are the most important ones for Flang, In reality there
are many more between Fortran and LLVM IR.)
```

Author: Done.

  (ptr <...> %.global_tid.,
   ptr <...> %.bound_tid.,
   ptr <...> %n,
   ptr <...> %b,
   ptr <...> %a) {
entry:
<...>
  call void @__kmpc_for_static_init_4(<...>)
<...>
omp.inner.for.body.i:
<...>
omp.loop.exit.i:
  call void @__kmpc_for_static_fini(<...>)
<...>
  ret void
}
```
(output edited for readability)

The body of `simple` no longer does all the work. Instead it uses
`__kmpc_fork_call` to tell the OpenMP
[runtime library](https://github.com/llvm/llvm-project/tree/main/openmp)
to run another function, `simple (.omp_outlined)` to do the work.

This second function is referred to as a "micro task". The runtime library
splits the work across many instances of the micro task and each time
the micro task function is called, it gets a different slice of the work.

The number of instances is only known at runtime, and can be controlled with
settings such as [`OMP_NUM_THREADS`](https://www.openmp.org/spec-html/5.0/openmpse50.html).

The LLVM IR representation of `@simple` includes labels like `omp.loop.exit.i`, but
these are not specific to OpenMP. They are just normal LLVM IR labels whose name
includes `omp`.

## Sharing Clang's OpenMP Knowledge

> In April 2019 LLVM Flang ("F18" at the time) was approved to join the LLVM
> Project. A month later, it was proposed that LLVM Flang should
> share OpenMP support code with Flang.

Reviewer 4: This feels kind of redundant with the previous text in mind. Perhaps something like
```suggestion
Shortly after the time that flang was approved to join the LLVM Project
(in April 2019), it was proposed that LLVM Flang should
share OpenMP support code with Flang.
```

Author: I've done a version of this, also I realised it needs to say Flang shares with Clang not Flang with Flang.

called the "frontend" driver, despite being behind the scenes.

In Flang's case the "compiler driver" is `flang` and the "frontend driver" is
`flang -fc1` (they are two separate tools contained in the same binary).

They are separate tools so that the compiler driver can provide an interface
suited to compiler users, with stable options that do not change over time.
While the frontend driver is suited to compiler developers, exposes internal
compiler details and does not have a stable set of options.

You can see the differences if you add `-###` to the compiler command:
```
$ ./bin/flang /tmp/test.f90 -march=armv8a+memtag -###
 "<...>/flang" "-fc1"
   "-triple" "aarch64-unknown-linux-gnu"
   "-target-feature" "+v8a"
   "-target-feature" "+mte"
 "/usr/bin/ld" \
   "-o" "a.out"
   "-L/usr/lib/gcc/aarch64-linux-gnu/11"
```
(output edited for readability)

The compiler driver has split the compilation into a job for the frontend
(`flang -fc1`) and the linker (`ld`). `-march=` has been converted into many
arguments to `flang -fc1`. This means that if compiler developers decided to
change how `-march=` was converted, existing compile commands for `flang` would
still work.

> Another compiler driver responsibility is to know where to find libraries on various

Reviewer 6: This is very much a matter of style and is, obviously, more verbose. 

```suggestion
Another responsibility of the compiler driver is to know where to find libraries on various
```

systems. This differs between operating systems and even distributions of the
same family of operating systems (for example Linux distributions).

This created a problem when implementing the compiler driver for Flang. This many
details would take a long time to get right.

> Luckily for Flang, by this time it was in the LLVM Project alongside Clang.

Reviewer 6: Not sure if the comma is strictly necessary, but it feels that way to me.

```suggestion
Luckily for Flang, by this time, it was in the LLVM Project alongside Clang.
```

Clang already knew how to handle this and had been tested on all sorts of
systems over many years.

> The intent is to mirror clang, for both the driver and CompilerInvocation, as
> much as makes sense to do so. The aim is to avoid re-inventing the wheel and
> to enable people who have worked with either the clang or flang entrypoints,
> drivers, and frontends to easily understand the other.
* [Peter Waller](https://discourse.llvm.org/t/rfc-adding-a-fortran-mode-to-the-clang-driver-for-flang/52307) (Arm)

> Flang became the the first in tree project to use Clang's compiler driver

Reviewer 6: ```suggestion
Flang became the the first in-tree project to use Clang's compiler driver
```

library (`clangDriver`) to implement its own compiler driver.

> This meant that Flang able to handle all the targets and tools that Clang could, without duplicating large amounts of code.

Reviewer 6: ```suggestion
This meant that Flang was able to handle all the targets and tools that Clang could, without duplicating large amounts of code.
```

# Reflections on Flang

We are almost 10 years from the first announcement of what would become LLVM

> Flang. In LLVM monorepo alone there have been close to 10,000 commits

Reviewer 6: ```suggestion
Flang. In the LLVM monorepo alone there have been close to 10,000 commits
```

from around 400 different contributors. Undoubtedly more in Classic Flang before
that.

> So it is time to hear from users, contributors, and supporters past and present

Reviewer 6: ```suggestion
So it is time to hear from users, contributors, and supporters, past and present,
```

  that is more palatable to commercial users and contributors.
* Develop an active community of Fortran compiler developers that includes
  companies and institutions.
* Support Fortran tool development by basing Flang on existing LLVM frameworks.
* Support Fortran language experimentation for future language standards proposals.

Intentions which were echoed by Pat McCormick:

> The overarching goal was to establish an open-source, modern implementation and
> simultaneously grow a community that spanned industry, academia, and federal
> agencies at both the national and international levels.

The utility of competing implementations may not be obvious. One advantage is
spreading out risks such as lack of funding, developer distribution and project
mismanagement.

Another is that it benefits Fortran's evolution as a language. For C++ language
features, it is common to implement them on top of Clang, GCC, or both, to
prove the feature is viable and gather feedback.

Implementing the feature multiple times in different compilers uncovers
assumptions that may be a problem for one compiler or another, or tradeoffs that

> that affect certain types of compiler user differently.

Reviewer 5: double "that"

Author: Fixed, thanks!
